sdk/Sdk.Generators/FunctionMetadataProviderGenerator.cs

```diff
@@ -0,0 +1,384 @@
+            indentedTextWriter.Indent++;
+            indentedTextWriter.WriteLine("Anonymous,");
+            indentedTextWriter.WriteLine("User,");
+            indentedTextWriter.WriteLine("Function,");
+            indentedTextWriter.WriteLine("System,");
+            indentedTextWriter.WriteLine("Admin");
+            indentedTextWriter.Indent--;
+            indentedTextWriter.WriteLine("}");
+        }
+
+        // adds a generated registration extension that users can call to register the source-generated function metadata provider
+        private static void AddRegistrationExtension(IndentedTextWriter indentedTextWriter)
+        {
+            indentedTextWriter.WriteLine("public static class WorkerHostBuilderFunctionMetadataProviderExtension");
+            indentedTextWriter.WriteLine("{");
+            indentedTextWriter.Indent++;
+            indentedTextWriter.WriteLine("public static IHostBuilder ConfigureGeneratedFunctionMetadataProvider(this IHostBuilder builder)");
+            indentedTextWriter.WriteLine("{");
+            indentedTextWriter.Indent++;
+            indentedTextWriter.WriteLine("builder.ConfigureServices(s => ");
+            indentedTextWriter.WriteLine("{");
+            indentedTextWriter.Indent++;
+            indentedTextWriter.WriteLine("s.AddSingleton<IFunctionMetadataProvider, GeneratedFunctionMetadataProvider>();");
+            indentedTextWriter.Indent--;
+            indentedTextWriter.WriteLine("});");
+            indentedTextWriter.WriteLine("return builder;");
+            indentedTextWriter.Indent--;
+            indentedTextWriter.WriteLine("}");
+            indentedTextWriter.Indent--;
+            indentedTextWriter.WriteLine("}");
+        }
+
+        /// <summary>
+        /// Register a factory that can create our custom syntax receiver
+        /// </summary>
+        /// <param name="context"></param>
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            context.RegisterForSyntaxNotifications(() => new SyntaxReceiver());
+        }
+
+        /// <summary>
+        /// Created on demand before each generation pass
+        /// </summary>
+        private class SyntaxReceiver : ISyntaxReceiver
+        {
+            public List<MethodDeclarationSyntax> CandidateMethods { get; } = new List<MethodDeclarationSyntax>();
+
+            /// <summary>
+            /// Called for every syntax node in the compilation, we can inspect the nodes and save any information useful for generation
+            /// </summary>
+            public void OnVisitSyntaxNode(SyntaxNode syntaxNode)
+            {
+                if (syntaxNode is MethodDeclarationSyntax methodSyntax)
+                {
+                    foreach (var p in methodSyntax.ParameterList.Parameters)
+                    {
+                        if (p.AttributeLists.Count > 0)
+                        {
+                            CandidateMethods.Add(methodSyntax);
```

Comment:
Are we collection the function methods here? Will this pick up any non function methods in the class? Should we check for the `Function` attribute existence to the if condition?

Reply:
This checks that there is at least one attribute in the method's parameters - checking for the Function attribute will likely miss cases like [this](https://github.com/Azure/azure-functions-dotnet-worker/blob/95a10c7aa95da2ba1122201ef78f5059733fa283/samples/FunctionApp/HttpTriggerWithBlobInput/HttpTriggerWithBlobInput.cs#L35) from our sample App. 

I do agree though that we might pick up methods that we shouldn't simply because they have an attribute - let me investigate further and see how we can design the `SyntaxReceiver` to work with all the scenarios we need.